sdk/lib/opentelemetry/sdk/trace/trace_state.rb

```diff
@@ -0,0 +1,186 @@
+        #
+        # @param [Hash<String, String>] hash Key-value pairs
+        # @return [TraceState]
+        def initialize(hash)
+          @hash = hash.freeze
+        end
+
+        # Returns the value associated with the given key, or nil if the key
+        # is not present.
+        #
+        # @param [String] key The key to lookup.
+        # @return [String] The value associated with the key, or nil.
+        def value(key)
+          @hash[key]
+        end
+
+        alias [] value
+
+        # Adds a new key/value pair or updates an existing value for a given key.
+        # Keys and values are validated against
+        # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-handling.md,
+        # and any invalid members are logged at DEBUG level and
+        # ignored. If adding the key/value pair would make the entire `ot` entry
+        # exceed the 256 character limit, the pair will be logged at DEBUG
+        # level and ignored.
+        #
+        # @param [String] key The key to add or update.
+        # @param [String] value The value to add or update.
+        # @return [TraceState] self, if unchanged, or a new TraceState containing
+        #   the new or updated key/value pair.
+        def set_value(key, value)
+          unless VALID_KEYS.include?(key) && VALID_VALUE.match?(value)
+            OpenTelemetry.logger.debug("Invalid Tracestate member - #{key} : #{value}")
+            return self
+          end
+
+          # TODO: length limit
+
+          h = Hash[@hash]
+          h[key] = value
+          self.class.create(h)
+        end
+
+        # Deletes the key/value pair associated with the given key.
+        #
+        # @param [String] key The key to remove.
+        # @return [TraceState] self, if unchanged, or a new TraceState without
+        #   the specified key.
+        def delete(key)
+          return self unless @hash.key?(key)
+
+          h = Hash[@hash]
+          h.delete(key)
+          self.class.create(h)
+        end
+
+        # Sets the value of the `ot` key in tracestate to this TraceState
+        # encoded as a string.
+        #
+        # @param [Tracestate] tracestate The Tracestate to update.
+        # @return [Tracestate] tracestate, if unchanged, or a new Tracestate
+        #   containing this TraceState under the `ot` key, encoded as a string.
+        def this_is_a_deliberately_terrible_name_please_bikeshed(tracestate)
+          tracestate.set_value('ot', to_s)
+        end
+
+        # Returns this TraceState encoded according to the specification at
+        # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-handling.md,
+        #
+        # @return [String] this TraceState encoded as a string.
+        def to_s
+          @hash.inject(+'') do |memo, (k, v)|
+            memo << k << ':' << v << ';'
+          end.chop! || ''
```

Comment:
Is the `+''` argument to `inject` something to do with frozen string literals? I've never seen that before.

While I'm at it, another way to do this might be:

```
irb(main):001:0> h = { 'foo' => 'bar', 'baz' => 'bat' }
=> {"foo"=>"bar", "baz"=>"bat"}
irb(main):002:0> h.map { _1.join(':') }.join(';')
=> "foo:bar;baz:bat"
irb(main):003:0> RUBY_VERSION
=> "2.7.2"
irb(main):004:0>
```

But it's a style thing I guess.

Reply:
https://ruby-doc.org/core-3.1.2/String.html#method-i-2B-40

> +string → new_string or self
> Returns `self` if `self` is not frozen.
> 
> Otherwise. returns `self.dup`, which is not frozen.

This gives us a mutable string, whereas string literals are usually frozen.

Reply:
The alternative style `h.map { _1.join(':') }.join(';')` will (probably) allocate more temporary strings if there are many keys. It's unlikely to matter much in practice, but this mimics what we do in the "other" `Tracestate` class, which is where I stole it from: https://github.com/open-telemetry/opentelemetry-ruby/blob/672739120a605c5c14afc48e5aa07d1320443355/api/lib/opentelemetry/trace/tracestate.rb#L127-L131

Reply:
Interesting - I went looking for that in the docs but must have overlooked it entirely.

Regarding the alternate style - if you suspect it'll allocate more, than never mind. Like I just commented elsewhere it _is_ important for us to be miserly with our allocations so I definitely prefer the less-expensive version. And, consistency with how we do it elsewhere (as you also said) is super good 👍. 😄